New transaction - PayingForTx

hanssv opened the pull request:
Allows an account to pay for a transaction on behalf of someone else.
Paying for fees and gas cost, will not cover the amount spent by the
transaction, just "the cost of the transaction" (and the extra size
added by wrapping the original transaction).

Note: this changes how "gas" for transaction size is calculated for GAs
from IRIS onwards. Instead of the outer/wrapping TX paying the size-gas
the whole TX it just has to cover its own part of the TX. This should be
less confusing, and easier to reason about. 

Closes #2851

hanssv commented:
Associated documentation https://github.com/aeternity/protocol/pull/443

ThomasArts commented on apps/aechannel/src/aesc_utils.erl:
```diff
@@ -734,30 +752,30 @@ process_solo_close(ChannelPubKey, FromPubKey, Nonce, Fee,
     add_event(
       process_solo_close_slash(ChannelPubKey, FromPubKey, Nonce, Fee,
-                               Payload, PoI, Height, Trees),
+                               Payload, PoI, Height, Trees, Env),
```
Height is already part of Env, but that optimization is not part of this PR.

ThomasArts commented on apps/aecore/src/aec_governance.erl:
```diff
@@ -158,7 +158,8 @@ tx_base_gas(name_update_tx, _)            -> ?TX_BASE_GAS;
 tx_base_gas(oracle_register_tx, _)        -> ?TX_BASE_GAS;
-tx_base_gas(oracle_response_tx, _)        -> ?TX_BASE_GAS.
+tx_base_gas(oracle_response_tx, _)        -> ?TX_BASE_GAS;
+tx_base_gas(paying_for_tx, _)             -> ?TX_BASE_GAS div 5.
```
Somewhere we should record the ratio behind this 1/5th.

hanssv replied:
How about here: https://github.com/aeternity/protocol/pull/443 (you have already approved it!)

tolbrino replied:
There should be an inline comment as well.

ThomasArts commented on apps/aecore/src/aec_paying_for_tx.erl:
```diff
@@ -0,0 +1,213 @@
+    {Type, InnerTx} = aetx:specialize_type(Tx),
+    case Type of
+        paying_for_tx                   -> {error, not_allowed_to_pay_for_recursively};
+        ga_meta_tx                      -> check_tx(aega_meta_tx:tx(InnerTx));
```
May the ga_meta_tx in its turn contain a paying_for_tx ? Then there is some indirect recursion.

hanssv replied:
No, that is exactly what we are checking here...

ThomasArts commented on apps/aega/test/aega_SUITE.erl:
```diff
@@ -418,17 +434,23 @@ basic_minimum_fee(_Cfg) ->
     SizeFee = aetx:size(DummyTx) * aec_governance:byte_gas() * MinGP,
-    MinimumInnerFee = MinimumSpendFee - SizeFee,
+    MinimumInnerFee = case Protocol =< ?LIMA_PROTOCOL_VSN of
+                        true  -> MinimumSpendFee - SizeFee;
```
This is not new code, but just wonder whether this could ever become a negative MinimumInnerFee?

hanssv replied:
No it can't. Since `SizeFee` is part of `MinimumSpendFee`.

ThomasArts commented on apps/aeprimop/src/aeprimop.erl:
```diff
@@ -215,7 +216,7 @@ oracle_extend_tx_instructions(Pubkey, DeltaTTL, Fee, Nonce) ->
                              QueryFee, QTTL, RTTL, TxFee, Nonce) ->
     [ force_inc_account_nonce_op(SenderPubkey, Nonce)
-    , spend_fee_op(SenderPubkey, TxFee + QueryFee)
+    , spend_fee_op(SenderPubkey, TxFee, QueryFee)
```
Ah, Query fee still has to be paid by inner Tx

hanssv replied:
Yes, the `QueryFee` does not go to the miner so it should be payed by the TX-actor.

ThomasArts commented on apps/aeprimop/src/aeprimop.erl:
```diff
@@ -566,13 +578,31 @@ lock_namefee(Kind, From, Amount, #state{protocol = Protocol} = S) ->
+    {spend_fee, {From, Amount, 0}}.
+
+spend_fee_op(From, DelegatedAmount, Amount)
+        when ?IS_HASH(From), ?IS_NON_NEG_INTEGER(DelegatedAmount + Amount) ->
```
Actually, one would like to enforce the stronger ?IS_NON_NEG_INTEGER(DelegateAmount), ?IS_NON_NEG_INTEGER(Amount).

hanssv replied:
Yes 👍

ThomasArts commented on apps/aeprimop/src/aeprimop.erl:
```diff
@@ -1122,17 +1164,20 @@ ga_attach({OwnerPubkey, GasLimit, GasPrice, ABIVersion,
     %% and the deposit (stored in the contract) to the contract owner (caller),
     CTVersion      = #{vm => VMVersion, abi => ABIVersion},
-    S2             = account_spend(Account, TotalAmount, S1),
+    S2             = if OwnerAmount > 0 -> account_spend(Account, OwnerAmount, S1);
```
I have a hard time to see when the OwnerAmount == 0 kicks in? Let alone <0.

hanssv replied:
OwnerAmount can't be `< 0` and if it is `0` the spending `0` is a no-op so we skip that.

ThomasArts commented:
A future feature request is to disable to possibilty to take the inner transaction and post it. As it is now, someone can promisse to pay for a transaction, but if the signer of that transaction actually has an account with tokens, the inner transaction can just be posted and inner transaction signer is charged for fee and gas.

hanssv commented:
Fixed Thomas' comment on asserts in `aeprimop` and rebased.

Did nothing about the potential problem that the inner TX is a valid transaction, and confirmed that `aetx_sign:serialize_for_client_inner` is correct for `aec_paying_for_tx`.

hanssv commented:
We have a plan for making the inner transaction invalid (in the post to mempool sense) by signing something slightly different (probably dropping the network id or similar). But let's postpone this to a separate PR/Issue.
